analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.IsNull.cs

```diff
@@ -0,0 +1,208 @@
+            validator.TagValues("UnknownToNotNull").Should().HaveCount(2)
+                .And.ContainSingle(x => x == null)
+                .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
+        }
+
+        [TestMethod]
+        public void IsNull_Coalesce_UnknownToUnknown()
+        {
+            const string code = @"
+var unknownToUnknown = arg ?? arg;
+Tag(""UnknownToUnknown"", unknownToUnknown);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateContainsOperation(OperationKind.IsNull);
+            validator.TagValues("UnknownToUnknown").Should().HaveCount(1)
+                .And.ContainSingle(x => x == null);
+        }
+
+        [TestMethod]
+        public void IsNull_CoalesceAssignment_SetsObjectConstraint()
+        {
+            const string code = @"
+object nullValue = null;
+object notNullValue = new object();
+var nullToNull = nullValue;     // Initial values for coalesce assignment
+var nullToNotNull = nullValue;
+var nullToUnknown = nullValue;
+var notNullToNull = notNullValue;
+var notNullToNotNull = notNullValue;
+var notNullToUnknown = notNullValue;
+nullToNull ??= nullValue;
+nullToNotNull ??= notNullValue;
+nullToUnknown ??= arg;
+notNullToNull ??= nullValue;
+notNullToNotNull ??= notNullValue;
+notNullToUnknown ??= arg;
+Tag(""NullToNull"", nullToNull);
+Tag(""NullToNotNull"", nullToNotNull);
+Tag(""NullToUnknown"", nullToUnknown);
+Tag(""NotNullToNull"", notNullToNull);
+Tag(""NotNullToNotNull"", notNullToNotNull);
+Tag(""NotNullToUnknown"", notNullToUnknown);
+";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateContainsOperation(OperationKind.IsNull);
+            validator.ValidateTag("NullToNull", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
+            validator.ValidateTag("NullToNotNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("NullToUnknown", x => x.Should().BeNull());
+            validator.ValidateTag("NotNullToNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("NotNullToNotNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("NotNullToUnknown", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+        }
+
+        [TestMethod]
+        public void IsNull_CoalesceAssignment_UnknownToNull()
+        {
+            const string code = @"
+object nullValue = null;
+arg ??= nullValue;
+Tag(""Arg"", arg);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateContainsOperation(OperationKind.IsNull);
+            validator.TagValues("Arg").Should().HaveCount(2)
+                .And.ContainSingle(x => x == null)
```

Comment:
What is this?

Reply:
`arg` is unknown while `nullValue` is known to be Null.
So IF `arg` is something not-null, the final result of ?? operation is still unknown. It's reassigned from `arg`.
ELSE it was null. And `null` was propagated from `nullValue`.

There's one more aspect of this - from the condition, we can learn that `arg` is actually once `null` and once `not-null`. These learning on the fly will be in another set of PRs. Especially for
```
if(arg == null)
 // Here, we know that arg is Null
else
  // Here, we know that arg is NotNull
```